### DISCUSSION TITLE
Quiet import error if import is wrapped in an import error exception handler

### DISCUSSION THREAD

**Discussion by rswdp** [microsoft/pylance-release#2702] [Enhancement]
## Environment data

Pylance 2020.6.1

## Expected behaviour

```python
try:
    from tabulate import tabulate
except ImportError:
    pass
```

Sometimes an optional third-party module may or may not be available, so you wrap an import in an exception handler that catches the import error and ignores it.  However, pylance presently displays an error for this when it should not.

## Actual behaviour

Error msg:
```
  Import "tabulate" could not be resolved from source
```

**Comment by erictraut** [comment]
This is intended behavior. You are correct that import statements can be included within try/except blocks or within conditional statements, but they are still import statements, and there is utility in being told when the import cannot be resolved. If you want to suppress this error message, you have a couple of options:

1. Disable that particular diagnostic for the file by adding a comment at the top:
```
# pyright: reportMissingModuleSource=false
```
2. Suppress the diagnostic for that particular line by adding a comment at the end of the line:
```
# type: ignore
```

**Comment by rswdp** [comment]
Ok, thanks for the quick response.  I will use #2 but I would rather have an option similar to #1 but that is triggered only when the language server determines that the ImportError is handled by the program since I would rather not mark up the program with annotations for language tools, though I understand knowing whether the error is handled or not may be complex to determine.

**Comment by savannahostrowski** [comment]
@rswdp Thanks for the feedback! Much appreciated. We've added the enhancement and needs decision tags while we do a little research on this experience.

**Comment by telamonian** [comment]
I have the exact same issue, and actually ended up opening a duplicate bug report over at #1005.

But yeah, :+1: from me to fixing nuisance warnings from the static checker for import of optional modules

**Comment by telamonian** [comment]
> This is intended behavior. You are correct that import statements can be included within try/except blocks or within conditional statements, but they are still import statements, and there is utility in being told when the import cannot be resolved.

@erictraut You're definitely right that there is still some utility in being told that the import cannot be resolved. But I think the current "intended behavior" misses a subtle distinction here about how that info should be treated, and how it should be reported to the user. To me, a warning reported at top level in vscode says "you, programmer, likely have a problem that needs fixing", which is not the case here.

Basically, I think that the way that unused imports are currently reported to the user (visually marked as distinct, more info revealed on hover) should be the model for handling this case. That way, the static analysis reported by vscode would better reflect the (non-problematic) optional module "concept" that is explicitly specified via a `try... except ImportError` clause

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
